### PR TITLE
feat(client): Runtime for model extensions

### PR DESCRIPTION
**Pull request by SevInf**
Core new idea here: we introduce new way of building our proxies from
building blocks. For example:

```
createCompositeProxy(target, [
  addProperty('foo', () => 1),
  forwardTo(someObject),
  modelLayer(client, modelName)
])
```

This will build a proxy around around `target`, that will add property
`foo` on top, all properties from `someObject` and all model methods
for `modelName`. Not only it allows us to reuse some of the code,
previously duplicated between different proxies, it will also makes
implementation of model and client extensions pretty straightforward:
each extension is just another layer of the composite proxy.

### CODE REVIEW

File: packages/client/tests/functional/extensions/enabled/model.ts [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+            firstMethod,
+          },
+        },
+      })
+      .$extends({
+        model: {
+          user: {
+            // TODO: remove any once types are generated
+            secondMethod(this: any) {
+              this.firstMethod()
```
Comment (millsp) [inline packages/client/tests/functional/extensions/enabled/model.ts:201]: I think this needs discussion as it is not currently permitted by the proposal.

File: packages/client/src/runtime/core/compositeProxy/createCompositeProxy.ts [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+  /**
+   * Returns a value for a property for a given key (one of the keys, returned
+   * from `getKeys`)
+   * @param key
+   */
+  getPropertyValue(key: KeyType): unknown
+
+  /**
+   * Gets a descriptor for given property. If not implemented or undefined is returned, { enumerable: true, writeable: true, configurable: true}
```
Comment (millsp) [inline packages/client/src/runtime/core/compositeProxy/createCompositeProxy.ts:17]:
```suggestion
   * Gets a descriptor for given property. If not implemented or undefined is returned, { enumerable: true, writeable: true, configurable: true} is defaulted
```

File: packages/client/src/runtime/core/compositeProxy/createCompositeProxy.ts [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    const keys = layer.getKeys()
+    for (const key of keys) {
+      keysToLayerMap.set(key, layer)
+    }
+  }
+  return keysToLayerMap
+}
+
+function getOwnKeys(target: object, layerKeys: (string | symbol)[]) {
+  const set = new Set([...Object.keys(target), ...layerKeys])
```
Comment (millsp) [inline packages/client/src/runtime/core/compositeProxy/createCompositeProxy.ts:92]:
```suggestion
return [...new Set([...Object.keys(target), ...layerKeys])]
```

File: packages/client/src/runtime/core/compositeProxy/forwardTo.ts [hunk 4]
```diff
@@ -0,0 +1,20 @@
+import type { CompositeProxyLayer } from './createCompositeProxy'
+
+/**
+ * Composite proxy layer that forwards all reads
+ * to provided object
+ *
+ * @param object
+ * @returns
+ */
+export function forwardTo(object: object): CompositeProxyLayer {
```
Comment (millsp) [inline packages/client/src/runtime/core/compositeProxy/forwardTo.ts:10]: I would prefer a different name like objectToCompositeProxyLayer. Much longer put more descriptive imo.

File: packages/client/src/runtime/core/model/applyModel.ts [hunk 5]
```diff
@@ -93,35 +110,38 @@ export function applyModel(client: Client, dmmfModelName: string) {
...
-  action: string,
-): action is Action | ClientOnlyModelAction {
-  if (isClientOnlyAction(action)) {
-    return isValidActionName(client, dmmfModelName, clientOnlyActions[action].wrappedAction)
-  }
-  return getOwnKeys(client, dmmfModelName).includes(action)
+  actionKeys.push('count')
+
+  const clientOnlyActionKeys = Object.keys(clientOnlyActions).filter((actionKey) =>
+    actionKeys.includes(clientOnlyActions[actionKey].wrappedAction),
```
Comment (millsp) [inline packages/client/src/runtime/core/model/applyModel.ts:130]: What is the reason for this change?

**Review by millsp — APPROVED** [review]
Looks great and thorough. Let's sync over the next steps tomorrow.

File: packages/client/src/runtime/core/model/applyModel.ts (hunk 5 above)
Comment (SevInf) [reply packages/client/src/runtime/core/model/applyModel.ts:130]: It think it was a mistake during inital `*OrThrow` method implementation: they were never exposed as enumerable properties. Since now layer  does not gets called if it does not declare the specific key, error became more obvious.

File: packages/client/src/runtime/core/compositeProxy/forwardTo.ts (hunk 4 above)
Comment (SevInf) [reply packages/client/src/runtime/core/compositeProxy/forwardTo.ts]: Renamed to `addObjectProperties`: more descriptive than `forwardTo`, but also consistent with `addProperty` layer. Let me know if you still prefer different name.

Comment (millsp) [reply packages/client/src/runtime/core/compositeProxy/forwardTo.ts]: That's better, thanks.
